subsys/fs/ext2/ext2_impl.c

```diff
@@ -0,0 +1,1542 @@
+/*
+ * Copyright (c) 2023 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/init.h>
+#include <zephyr/fs/fs.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/util.h>
+
+#include "ext2.h"
+#include "ext2_impl.h"
+#include "ext2_struct.h"
+#include "ext2_diskops.h"
+#include "ext2_bitmap.h"
+
+LOG_MODULE_REGISTER(ext2, CONFIG_EXT2_LOG_LEVEL);
+
+static struct ext2_data __fs;
+static bool initialized;
+
+/* Structures for blocks slab alocator */
+struct k_mem_slab ext2_block_memory_slab, ext2_block_struct_slab;
+char __aligned(8) __ext2_block_memory_buffer[CONFIG_EXT2_BLOCKS_BUF_SIZE];
+char __aligned(8) __ext2_block_struct_buffer[CONFIG_EXT2_BLOCKS_BUF_SIZE / 64];
+
+/* Initialize heap memory allocator */
+K_HEAP_DEFINE(ext2_heap, CONFIG_EXT2_HEAP_SIZE);
+
+/* Helper functions --------------------------------------------------------- */
+
+void error_beharior(struct ext2_data *fs, const char *msg)
+{
+	LOG_ERR("File system corrupted: %s", msg);
+
+	/* If file system is not initialized panic */
+	if (!fs->sblock) {
+		LOG_ERR("File system data not found. Panic...");
+		k_panic();
+	}
+
+	switch (EXT2_DATA_SBLOCK(fs)->s_errors) {
+	case EXT2_ERRORS_CONTINUE:
+		/* Do nothing */
+		break;
+	case EXT2_ERRORS_RO:
+		LOG_WRN("Marking file system as read only");
+		fs->flags |= EXT2_DATA_FLAGS_RO;
+		break;
+	case EXT2_ERRORS_PANIC:
+		LOG_ERR("Panic...");
+		k_panic();
+		break;
+	default:
+		LOG_ERR("Unrecognized errors behavior in superblock s_errors field. Panic...");
+		k_panic();
+	}
+}
+
+void *ext2_heap_alloc(size_t size)
+{
+	void *ptr = k_heap_alloc(&ext2_heap, size, K_NO_WAIT);
+
+	if (!ptr) {
+		return NULL;
+	}
+
+	memset(ptr, 0, size);
```

Comment:
Do you want to really do that here? I can already see a few instances where entire structure is immediately filled in after being allocated, which means this is wasted `memset` call. You save on flash moving this here and applying to every allocated buffer, but that takes toll on CPU, often, for no reason.

Reply:
I've moved it out to the places where it is truly needed.